pycbc/workflow/configuration.py

```diff
@@ -29,65 +29,183 @@
-       url.startswith('file://'):
-        filename = url.split('/')[-1]
-        filename = os.path.join(directory, filename)
-        succeeded = False
-        num_tries = 5
-        t_sleep   = 10
-
-        while not succeeded and num_tries > 0: 
-            try:
-                response = urllib2.urlopen(url)
-                result   = response.read()
-                out_file = open(filename, 'w')
-                out_file.write(result)
-                out_file.close()
-                succeeded = True
-            except:
-                logging.warn("Unable to download %s, retrying" % url)
-                time.sleep(t_sleep)
-                num_tries -= 1
-                t_sleep   *= 2
-                
-        if not succeeded:
-            errMsg  = "Unable to download %s " % (url)
-            raise ValueError(errMsg)
-
-    elif url.find('://') != -1:
+    filename = os.path.join(directory,os.path.basename(u.path))
+
+    if u.scheme == '' or u.scheme == 'file':
+        # for regular files, make a direct copy
+        if os.path.isfile(u.path):
+            shutil.copy(u.path,filename)
+        else:
+            errmsg  = "Cannot open file %s from URL %s" % (u.path, url)
+            raise ValueError(errmsg)
+
+    elif u.scheme == 'http' or u.scheme == 'https':
+        s = requests.Session()
+        s.mount(str(u.scheme)+'://', 
+            requests.adapters.HTTPAdapter(max_retries=5))
+
+        # look for an ecp cookie file and load the cookies
+        cookie_dict = {}
+        ecp_file = '/tmp/ecpcookie.u%d' % os.getuid()
+        if os.path.isfile(ecp_file):
+            cj = cookielib.MozillaCookieJar()
+            cj.load(ecp_file, ignore_discard=True, ignore_expires=True)
+        else:
+            cj = []
+
+        for c in cj: 
+            if c.domain == u.netloc:
+                # load cookies for this server
+                cookie_dict[c.name] = c.value
+            elif u.netloc == "code.pycbc.phy.syr.edu" and \
+              c.domain == "git.ligo.org":
+                # handle the redirect for code.pycbc to git.ligo.org
+                cookie_dict[c.name] = c.value
+
+        r = s.get(url, cookies=cookie_dict, allow_redirects=True)
+        if r.status_code != 200:
+            errmsg = "Unable to download %s\nError code = %d" % (url,
+                r.status_code)
+            raise ValueError(errmsg)
+
+        # if we are downloading from git.ligo.org, check that we
+        # did not get redirected to the sign-in page
+        if u.netloc == 'git.ligo.org' or u.netloc == 'code.pycbc.phy.syr.edu':
+            soup = BeautifulSoup(r.content, 'html.parser')
+            desc = soup.findAll(attrs={"property":"og:url"})
+            if len(desc) and \
+              desc[0]['content'] == 'https://git.ligo.org/users/sign_in':
+                errmsg = "The attempt to download the file at\n\n%s\n" % url
+                errmsg += """
```

Comment:
I would prefer if multi-line error messages are declared above the function, and then used with the string.format() (or string % (args), although that is now unfavoured) in line.